daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Upgrading.scala

```diff
@@ -0,0 +1,464 @@
+      case _ => None
+    }
+  }
+
+}
+
+case class TypecheckUpgrades(packagesAndIds: Upgrading[(Ref.PackageId, Ast.Package)]) {
+  private lazy val packageId: Upgrading[Ref.PackageId] = packagesAndIds.map(_._1)
+  private lazy val _package: Upgrading[Ast.Package] = packagesAndIds.map(_._2)
+
+  private def extractDelExistNew[K, V](
+      arg: Upgrading[Map[K, V]]
+  ): (Map[K, V], Map[K, Upgrading[V]], Map[K, V]) =
+    (
+      arg.past -- arg.present.keySet,
+      arg.past.keySet
+        .intersect(arg.present.keySet)
+        .map(k => k -> Upgrading(arg.past(k), arg.present(k)))
+        .toMap,
+      arg.present -- arg.past.keySet,
+    )
+
+  private def checkDeleted[K, V, T <: Throwable](
+      arg: Upgrading[Map[K, V]],
+      handler: (K, V) => T,
+  ): Try[(Map[K, Upgrading[V]], Map[K, V])] = {
+    val (deletedV, existingV, newV) = extractDelExistNew(arg)
+    deletedV.headOption match {
+      case Some((k, v)) => Failure(handler(k, v))
+      case _ => Success((existingV, newV))
+    }
+  }
+
+  private def tryAll[A, B](t: Iterable[A], f: A => Try[B]): Try[Seq[B]] =
+    Try(t.map(f(_).get).toSeq)
+
+  private def check(): Try[Unit] = {
+    for {
+      (upgradedModules, newModules @ _) <-
+        checkDeleted(
+          _package.map(_.modules),
+          (name: Ref.ModuleName, _: Ast.Module) => MissingModule(name),
+        )
+      _ <- tryAll(upgradedModules.values, checkModule(_))
+    } yield ()
+  }
+
+  private def checkModule(module: Upgrading[Ast.Module]): Try[Unit] = {
+    def datatypes(module: Ast.Module): Map[Ref.DottedName, Ast.DDataType] =
+      module.definitions.flatMap(_ match {
+        case (k, v: Ast.DDataType) => Some((k, v));
+        case _ => None;
+      })
+
+    val moduleWithMetadata = module.map(ModuleWithMetadata)
+    for {
+      (existingTemplates, _new) <- checkDeleted(
+        module.map(_.templates),
+        (name: Ref.DottedName, _: Ast.Template) => MissingTemplate(name),
+      )
+      _ <- tryAll(existingTemplates, checkTemplate(_))
+
+      (existingDatatypes, _new) <- checkDeleted(
+        module.map(datatypes(_)),
+        (name: Ref.DottedName, _: Ast.DDataType) => MissingDataCon(name),
+      )
+      _ <- tryAll(existingDatatypes, checkDatatype(moduleWithMetadata, _))
+    } yield ()
+  }
+
+  private def checkTemplate(
+      templateAndName: (Ref.DottedName, Upgrading[Ast.Template])
+  ): Try[Unit] = {
+    val (templateName, template) = templateAndName
+    for {
+      (existingChoices, _newChoices) <- checkDeleted(
+        template.map(_.choices),
+        (name: Ref.ChoiceName, _: Ast.TemplateChoice) => MissingChoice(name),
+      )
+
+      _ <- tryAll(existingChoices.values, checkChoice(_))
+      _ <- checkKey(templateName, template.map(_.key))
+    } yield ()
+  }
+
+  private def checkType(typ: Upgrading[Ast.Type]): Boolean = {
+    AlphaEquiv.alphaEquiv(unifyPackageIds(typ.past), unifyPackageIds(typ.present))
```

Comment:
You shouldn't need to unify the `typ.past` unless you change it to check multiple package ids, since present should be able to exist in past.

Reply:
Assuming you mean "present should*n't* be able to exist in past" - good point, but Carl's PR is already using this logic, so I'll keep it this way just to spare him more work in the rebase, since the change is essentially a no-op for now.